rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,849 @@
+---
+authors: Alan Parra (alan.parra@goteleport.com), Erik Tate (erik.tate@goteleport.com)
+state: in development
+---
+
+# RFD 127 - Encrypted Session Recordings
+
+## Required Approvers
+
+- Engineering: @rosstimothy, @zmb3, @espadolini, @nklaassen
+- Security: doyensec
+
+## What
+
+This document proposes an approach to encrypting session recording data before
+writing to disk or any long term storage.
+
+## Why
+
+Recordings temporarily stored to disk can be easily tampered with by users with
+enough access. This could even occur within the session being recorded if the
+host user has root access.
+
+Encrypting session recordings at rest can help prevent exposure of credentials
+or other secrets that might be visible within the recording.
+
+## Details
+
+This document should fulfill the following requirements:
+
+- Encrypting session recording data at rest in long term storage and during any
+  intermediate disk writes.
+- Replay of encrypted sessions using the web UI and `tsh`.
+- Support sourcing key material directly from an HSM or other supported
+  keystore.
+- All auth servers in a cluster should be able to replay session recordings
+  even if they did not encrypt them.
+- An encryption algorithm suitable for this workload.
+
+### Encryption Algorithm
+
+This document proposes [age](https://github.com/FiloSottile/age) as the
+encryption format for session recordings. It was chosen for its provenance,
+simplicity, and focus on strong cryptography defaults without requiring
+customization. The formal spec can be found
+[here](https://age-encryption.org/v1). The officially supported key wrapping
+algorithms for `age` are limited to X25519, Ed25519, and RSA. Support for other
+algorithms requires implementing a custom plugin or requesting modifications
+from the upstream. The algorithms employed by `age` are not currently
+compatible with FIPS, which means configuring encrypted sessions while in FIPS
+mode will result in failed startup.
+
+In order to support key unwrapping directly within HSM and KMS systems, this
+RFD proposes a custom `age` plugin that uses RSA 4096 keypairs in combination
```

Comment:
Are we going with RSA-4096 for some specific security level that the more common RSA-2048 does not provide? If so, please mention it.

Reply:
We are, 128-bit is the minimum. I've updated the RFD to mention that :+1: